// mongodb/src/index.ts
import { Admin } from 'mongodb/src/admin';
import { ObjectId } from 'mongodb/src/bson';
import { Collection } from 'mongodb/src/collection';
import { AbstractCursor } from 'mongodb/src/cursor/abstract_cursor';
import { AggregationCursor } from 'mongodb/src/cursor/aggregation_cursor';
import { FindCursor } from 'mongodb/src/cursor/find_cursor';
import { ListCollectionsCursor } from 'mongodb/src/cursor/list_collections_cursor';
import { ListIndexesCursor } from 'mongodb/src/cursor/list_indexes_cursor';
import { Db } from 'mongodb/src/db';
import { GridFSBucket } from 'mongodb/src/gridfs';
import { Logger } from 'mongodb/src/logger';
import { MongoClient } from 'mongodb/src/mongo_client';
import { CancellationToken } from 'mongodb/src/mongo_types';
import { PromiseProvider } from 'mongodb/src/promise_provider';

export {
  Binary,
  BSONRegExp,
  BSONSymbol,
  Code,
  DBRef,
  Decimal128,
  Double,
  Int32,
  Long,
  Map,
  MaxKey,
  MinKey,
  ObjectId,
  Timestamp
} from 'mongodb/src/bson';
export { ChangeStreamCursor } from 'mongodb/src/cursor/change_stream_cursor';
/**
 * @public
 * @deprecated Please use `ObjectId`
 */
export const ObjectID = ObjectId;

export { AnyBulkWriteOperation, BulkWriteOptions, MongoBulkWriteError } from 'mongodb/src/bulk/common';
export {
  MongoAPIError,
  MongoAWSError,
  MongoBatchReExecutionError,
  MongoChangeStreamError,
  MongoCompatibilityError,
  MongoCursorExhaustedError,
  MongoCursorInUseError,
  MongoDecompressionError,
  MongoDriverError,
  MongoError,
  MongoExpiredSessionError,
  MongoGridFSChunkError,
  MongoGridFSStreamError,
  MongoInvalidArgumentError,
  MongoKerberosError,
  MongoMissingCredentialsError,
  MongoMissingDependencyError,
  MongoNetworkError,
  MongoNetworkTimeoutError,
  MongoNotConnectedError,
  MongoParseError,
  MongoRuntimeError,
  MongoServerClosedError,
  MongoServerError,
  MongoServerSelectionError,
  MongoSystemError,
  MongoTailableCursorError,
  MongoTopologyClosedError,
  MongoTransactionError,
  MongoUnexpectedServerResponseError,
  MongoWriteConcernError
} from 'mongodb/src/error';
export {
  AbstractCursor,
  // Actual driver classes exported
  Admin,
  AggregationCursor,
  CancellationToken,
  Collection,
  Db,
  FindCursor,
  GridFSBucket,
  ListCollectionsCursor,
  ListIndexesCursor,
  Logger,
  MongoClient,
  // Utils
  PromiseProvider as Promise
};

// enums
export { BatchType } from 'mongodb/src/bulk/common';
export { GSSAPICanonicalizationValue } from 'mongodb/src/cmap/auth/gssapi';
export { AuthMechanism } from 'mongodb/src/cmap/auth/providers';
export { Compressor } from 'mongodb/src/cmap/wire_protocol/compression';
export { CURSOR_FLAGS } from 'mongodb/src/cursor/abstract_cursor';
export { AutoEncryptionLoggerLevel } from 'mongodb/src/deps';
export { MongoErrorLabel } from 'mongodb/src/error';
export { ExplainVerbosity } from 'mongodb/src/explain';
export { LoggerLevel } from 'mongodb/src/logger';
export { ServerApiVersion } from 'mongodb/src/mongo_client';
export { BSONType } from 'mongodb/src/mongo_types';
export { ReturnDocument } from 'mongodb/src/operations/find_and_modify';
export { ProfilingLevel } from 'mongodb/src/operations/set_profiling_level';
export { ReadConcernLevel } from 'mongodb/src/read_concern';
export { ReadPreferenceMode } from 'mongodb/src/read_preference';
export { ServerType, TopologyType } from 'mongodb/src/sdam/common';

// Helper classes
export { ReadConcern } from 'mongodb/src/read_concern';
export { ReadPreference } from 'mongodb/src/read_preference';
export { WriteConcern } from 'mongodb/src/write_concern';

// events
export {
  CommandFailedEvent,
  CommandStartedEvent,
  CommandSucceededEvent
} from 'mongodb/src/cmap/command_monitoring_events';
export {
  ConnectionCheckedInEvent,
  ConnectionCheckedOutEvent,
  ConnectionCheckOutFailedEvent,
  ConnectionCheckOutStartedEvent,
  ConnectionClosedEvent,
  ConnectionCreatedEvent,
  ConnectionPoolClearedEvent,
  ConnectionPoolClosedEvent,
  ConnectionPoolCreatedEvent,
  ConnectionPoolMonitoringEvent,
  ConnectionReadyEvent
} from 'mongodb/src/cmap/connection_pool_events';
export {
  ServerClosedEvent,
  ServerDescriptionChangedEvent,
  ServerHeartbeatFailedEvent,
  ServerHeartbeatStartedEvent,
  ServerHeartbeatSucceededEvent,
  ServerOpeningEvent,
  TopologyClosedEvent,
  TopologyDescriptionChangedEvent,
  TopologyOpeningEvent
} from 'mongodb/src/sdam/events';
export { SrvPollingEvent } from 'mongodb/src/sdam/srv_polling';

// type only exports below, these are removed from emitted JS
export type { AdminPrivate } from 'mongodb/src/admin';
export type { BSONSerializeOptions, Document } from 'mongodb/src/bson';
export type { deserialize, serialize } from 'mongodb/src/bson';
export type {
  BulkResult,
  BulkWriteOperationError,
  BulkWriteResult,
  DeleteManyModel,
  DeleteOneModel,
  InsertOneModel,
  ReplaceOneModel,
  UpdateManyModel,
  UpdateOneModel,
  WriteConcernError,
  WriteError
} from 'mongodb/src/bulk/common';
export type {
  Batch,
  BulkOperationBase,
  BulkOperationPrivate,
  FindOperators,
  WriteConcernErrorData
} from 'mongodb/src/bulk/common';
export type { OrderedBulkOperation } from 'mongodb/src/bulk/ordered';
export type { UnorderedBulkOperation } from 'mongodb/src/bulk/unordered';
export type {
  ChangeStream,
  ChangeStreamCollModDocument,
  ChangeStreamCreateDocument,
  ChangeStreamCreateIndexDocument,
  ChangeStreamDeleteDocument,
  ChangeStreamDocument,
  ChangeStreamDocumentCollectionUUID,
  ChangeStreamDocumentCommon,
  ChangeStreamDocumentKey,
  ChangeStreamDocumentOperationDescription,
  ChangeStreamDropDatabaseDocument,
  ChangeStreamDropDocument,
  ChangeStreamDropIndexDocument,
  ChangeStreamEvents,
  ChangeStreamInsertDocument,
  ChangeStreamInvalidateDocument,
  ChangeStreamNameSpace,
  ChangeStreamOptions,
  ChangeStreamRefineCollectionShardKeyDocument,
  ChangeStreamRenameDocument,
  ChangeStreamReplaceDocument,
  ChangeStreamReshardCollectionDocument,
  ChangeStreamShardCollectionDocument,
  ChangeStreamUpdateDocument,
  OperationTime,
  PipeOptions,
  ResumeOptions,
  ResumeToken,
  UpdateDescription
} from 'mongodb/src/change_stream';
export type {
  AuthMechanismProperties,
  MongoCredentials,
  MongoCredentialsOptions
} from 'mongodb/src/cmap/auth/mongo_credentials';
export type {
  BinMsg,
  MessageHeader,
  Msg,
  OpMsgOptions,
  OpQueryOptions,
  OpResponseOptions,
  Query,
  Response,
  WriteProtocolMessageType
} from 'mongodb/src/cmap/commands';
export type { LEGAL_TCP_SOCKET_OPTIONS, LEGAL_TLS_SOCKET_OPTIONS, Stream } from 'mongodb/src/cmap/connect';
export type {
  CommandOptions,
  Connection,
  ConnectionEvents,
  ConnectionOptions,
  DestroyOptions,
  GetMoreOptions,
  ProxyOptions,
  QueryOptions
} from 'mongodb/src/cmap/connection';
export type {
  CloseOptions,
  ConnectionPool,
  ConnectionPoolEvents,
  ConnectionPoolOptions,
  PoolState,
  WaitQueueMember,
  WithConnectionCallback
} from 'mongodb/src/cmap/connection_pool';
export type {
  MessageStream,
  MessageStreamOptions,
  OperationDescription
} from 'mongodb/src/cmap/message_stream';
export type { ConnectionPoolMetrics } from 'mongodb/src/cmap/metrics';
export type { StreamDescription, StreamDescriptionOptions } from 'mongodb/src/cmap/stream_description';
export type { CompressorName } from 'mongodb/src/cmap/wire_protocol/compression';
export type { CollectionOptions, CollectionPrivate, ModifyResult } from 'mongodb/src/collection';
export type { MONGO_CLIENT_EVENTS } from 'mongodb/src/constants';
export type {
  AbstractCursorEvents,
  AbstractCursorOptions,
  CursorCloseOptions,
  CursorFlag,
  CursorStreamOptions
} from 'mongodb/src/cursor/abstract_cursor';
export type { InternalAbstractCursorOptions } from 'mongodb/src/cursor/abstract_cursor';
export type { AggregationCursorOptions } from 'mongodb/src/cursor/aggregation_cursor';
export type {
  ChangeStreamAggregateRawResult,
  ChangeStreamCursorOptions
} from 'mongodb/src/cursor/change_stream_cursor';
export type { DbOptions, DbPrivate } from 'mongodb/src/db';
export type { AutoEncrypter, AutoEncryptionOptions, AutoEncryptionTlsOptions } from 'mongodb/src/deps';
export type { Encrypter, EncrypterOptions } from 'mongodb/src/encrypter';
export type { AnyError, ErrorDescription, MongoNetworkErrorOptions } from 'mongodb/src/error';
export type { Explain, ExplainOptions, ExplainVerbosityLike } from 'mongodb/src/explain';
export type {
  GridFSBucketReadStream,
  GridFSBucketReadStreamOptions,
  GridFSBucketReadStreamOptionsWithRevision,
  GridFSBucketReadStreamPrivate,
  GridFSFile
} from 'mongodb/src/gridfs/download';
export type { GridFSBucketEvents, GridFSBucketOptions, GridFSBucketPrivate } from 'mongodb/src/gridfs';
export type {
  GridFSBucketWriteStream,
  GridFSBucketWriteStreamOptions,
  GridFSChunk
} from 'mongodb/src/gridfs/upload';
export type { LoggerFunction, LoggerOptions } from 'mongodb/src/logger';
export type {
  Auth,
  DriverInfo,
  MongoClientEvents,
  MongoClientOptions,
  MongoClientPrivate,
  MongoOptions,
  PkFactory,
  ServerApi,
  SupportedNodeConnectionOptions,
  SupportedSocketOptions,
  SupportedTLSConnectionOptions,
  SupportedTLSSocketOptions,
  WithSessionCallback
} from 'mongodb/src/mongo_client';
export type {
  CommonEvents,
  EventsDescription,
  GenericListener,
  TypedEventEmitter
} from 'mongodb/src/mongo_types';
export type {
  AcceptedFields,
  AddToSetOperators,
  AlternativeType,
  ArrayElement,
  ArrayOperator,
  BitwiseFilter,
  BSONTypeAlias,
  Condition,
  EnhancedOmit,
  Filter,
  FilterOperations,
  FilterOperators,
  Flatten,
  InferIdType,
  IntegerType,
  IsAny,
  Join,
  KeysOfAType,
  KeysOfOtherType,
  MatchKeysAndValues,
  NestedPaths,
  NestedPathsOfType,
  NonObjectIdLikeDocument,
  NotAcceptedFields,
  NumericType,
  OneOrMore,
  OnlyFieldsOfType,
  OptionalId,
  OptionalUnlessRequiredId,
  Projection,
  ProjectionOperators,
  PropertyType,
  PullAllOperator,
  PullOperator,
  PushOperator,
  RegExpOrString,
  RootFilterOperators,
  SchemaMember,
  SetFields,
  UpdateFilter,
  WithId,
  WithoutId
} from 'mongodb/src/mongo_types';
export type { AddUserOptions, RoleSpecification } from 'mongodb/src/operations/add_user';
export type {
  AggregateOperation,
  AggregateOptions,
  DB_AGGREGATE_COLLECTION
} from 'mongodb/src/operations/aggregate';
export type {
  CollationOptions,
  CommandOperation,
  CommandOperationOptions,
  OperationParent
} from 'mongodb/src/operations/command';
export type { IndexInformationOptions } from 'mongodb/src/operations/common_functions';
export type { CountOptions } from 'mongodb/src/operations/count';
export type { CountDocumentsOptions } from 'mongodb/src/operations/count_documents';
export type {
  ClusteredCollectionOptions,
  CreateCollectionOptions,
  TimeSeriesCollectionOptions
} from 'mongodb/src/operations/create_collection';
export type { DeleteOptions, DeleteResult, DeleteStatement } from 'mongodb/src/operations/delete';
export type { DistinctOptions } from 'mongodb/src/operations/distinct';
export type { DropCollectionOptions, DropDatabaseOptions } from 'mongodb/src/operations/drop';
export type { EstimatedDocumentCountOptions } from 'mongodb/src/operations/estimated_document_count';
export type { EvalOptions } from 'mongodb/src/operations/eval';
export type { ExecutionResult } from 'mongodb/src/operations/execute_operation';
export type { FindOptions } from 'mongodb/src/operations/find';
export type {
  FindOneAndDeleteOptions,
  FindOneAndReplaceOptions,
  FindOneAndUpdateOptions
} from 'mongodb/src/operations/find_and_modify';
export type {
  CreateIndexesOptions,
  DropIndexesOptions,
  IndexDescription,
  IndexDirection,
  IndexSpecification,
  ListIndexesOptions
} from 'mongodb/src/operations/indexes';
export type { InsertManyResult, InsertOneOptions, InsertOneResult } from 'mongodb/src/operations/insert';
export type { CollectionInfo, ListCollectionsOptions } from 'mongodb/src/operations/list_collections';
export type { ListDatabasesOptions, ListDatabasesResult } from 'mongodb/src/operations/list_databases';
export type {
  FinalizeFunction,
  MapFunction,
  MapReduceOptions,
  ReduceFunction
} from 'mongodb/src/operations/map_reduce';
export type { AbstractOperation, Hint, OperationOptions } from 'mongodb/src/operations/operation';
export type { ProfilingLevelOptions } from 'mongodb/src/operations/profiling_level';
export type { RemoveUserOptions } from 'mongodb/src/operations/remove_user';
export type { RenameOptions } from 'mongodb/src/operations/rename';
export type { RunCommandOptions } from 'mongodb/src/operations/run_command';
export type { SetProfilingLevelOptions } from 'mongodb/src/operations/set_profiling_level';
export type {
  CollStats,
  CollStatsOptions,
  DbStatsOptions,
  WiredTigerData
} from 'mongodb/src/operations/stats';
export type {
  ReplaceOptions,
  UpdateOptions,
  UpdateResult,
  UpdateStatement
} from 'mongodb/src/operations/update';
export type { ValidateCollectionOptions } from 'mongodb/src/operations/validate_collection';
export type { ReadConcernLike } from 'mongodb/src/read_concern';
export type {
  HedgeOptions,
  ReadPreferenceFromOptions,
  ReadPreferenceLike,
  ReadPreferenceLikeOptions,
  ReadPreferenceOptions
} from 'mongodb/src/read_preference';
export type { ClusterTime, TimerQueue } from 'mongodb/src/sdam/common';
export type {
  Monitor,
  MonitorEvents,
  MonitorOptions,
  MonitorPrivate,
  RTTPinger,
  RTTPingerOptions
} from 'mongodb/src/sdam/monitor';
export type { Server, ServerEvents, ServerOptions, ServerPrivate } from 'mongodb/src/sdam/server';
export type {
  ServerDescription,
  ServerDescriptionOptions,
  TagSet,
  TopologyVersion
} from 'mongodb/src/sdam/server_description';
export type { ServerSelector } from 'mongodb/src/sdam/server_selection';
export type { SrvPoller, SrvPollerEvents, SrvPollerOptions } from 'mongodb/src/sdam/srv_polling';
export type {
  ConnectOptions,
  SelectServerOptions,
  ServerCapabilities,
  ServerSelectionCallback,
  ServerSelectionRequest,
  Topology,
  TopologyEvents,
  TopologyOptions,
  TopologyPrivate
} from 'mongodb/src/sdam/topology';
export type { TopologyDescription, TopologyDescriptionOptions } from 'mongodb/src/sdam/topology_description';
export type {
  ClientSession,
  ClientSessionEvents,
  ClientSessionOptions,
  EndSessionOptions,
  ServerSession,
  ServerSessionId,
  ServerSessionPool,
  WithTransactionCallback
} from 'mongodb/src/sessions';
export type { Sort, SortDirection, SortDirectionForCmd, SortForCmd } from 'mongodb/src/sort';
export type { Transaction, TransactionOptions, TxnState } from 'mongodb/src/transactions';
export type {
  BufferPool,
  Callback,
  ClientMetadata,
  ClientMetadataOptions,
  EventEmitterWithState,
  HostAddress,
  InterruptibleAsyncInterval,
  MongoDBNamespace
} from 'mongodb/src/utils';
export type { W, WriteConcernOptions, WriteConcernSettings } from 'mongodb/src/write_concern';
